### PR TITLE
Story/vspc 172

### PR DESCRIPTION
**Pull request by pradeep-017**
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

Description on hover for spacelink and modulelink.

...Put description here...
  
## Anything else the reviewer needs to know?

... describe here ...

### CODE REVIEW

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 1]
```diff
@@ -1466,6 +1478,15 @@ <h5 class="alert-heading">Create new Space Link</h5>
...
 	                </div>
+	                
+	                <div class="row">
+	                    <div class="col-sm-4">
+	                    	<label>Description</label>
+	                    </div>
+	                    <div class="col-sm-8">
+							<textarea class="form-control-sm target" rows="5" name="spaceLinkDesc" id="spaceLinkDesc" style="background: var(--base-grey); width: inherit;"></textarea>
+						</div>
+					</div>
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:1489]: indentation

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 2]
```diff
@@ -1550,6 +1571,15 @@ <h5 class="alert-heading">Create new Module Link</h5>
...
 	                </div>
+	                
+	                <div class="row">
+	                    <div class="col-sm-4">
+	                    	<label>Description</label>
+	                    </div>
+	                    <div class="col-sm-8">
+							<textarea class="form-control-sm modulelink-target" rows="5" name="moduleLinkDesc" id="moduleLinkDesc" style="background: var(--base-grey); width: inherit;"></textarea>
+						</div>
+					</div>
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:1582]: indentation

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 3]
```diff
@@ -1599,6 +1629,15 @@ <h5 class="alert-heading">Create new External Link</h5>
...
 	                </div>
+	                
+	                <div class="row">
+	                    <div class="col-sm-4">
+	                    	<label>Description</label>
+	                    </div>
+	                    <div class="col-sm-8">
+							<textarea class="form-control-sm extlink-target" rows="5" name="externalLinkDesc" id="externalLinkDesc" style="background: var(--base-grey); width: inherit;"></textarea>
+						</div>
+					</div>
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:1640]: indentation

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 4]
```diff
@@ -1686,6 +1725,15 @@ <h6 class="alert-heading"><span style="color: var(--dark-grey)"> Space Link:</sp
...
 	                </div>
+	                
+	                <div class="row">
+	                    <div class="col-sm-4">
+	                    	<label>Description</label>
+	                    </div>
+	                    <div class="col-sm-8">
+							<textarea class="form-control-sm spacelink-targetEdit" rows="5" name="spaceLinkDesc" id="spaceLinkDescEdit" style="background: var(--base-grey); width: inherit;"></textarea>
+						</div>
+					</div>
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:1736]: indentation

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 5]
```diff
@@ -1786,6 +1834,14 @@ <h6 class="alert-heading">
...
                         </div>
                     </div>
+                    <div class="row">
+	                    <div class="col-sm-4">
+	                    	<label>Description:</label>
+	                    </div>
+	                    <div class="col-sm-8">
+							<textarea class="form-control-sm modulelink-targetEdit" rows="5" name="moduleLinkDesc" id="moduleLinkDescEdit" style="background: var(--base-grey); width: inherit;"></textarea>
+						</div>
+					</div>
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:1844]: indentation

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 6]
```diff
@@ -1851,6 +1907,15 @@ <h6 class="alert-heading">
...
                     </div>
+                    
+                    <div class="row">
+	                    <div class="col-sm-4">
+	                    	<label>Description:</label>
+	                    </div>
+	                    <div class="col-sm-8">
+							<textarea class="form-control-sm externallink-targetEdit" rows="5" name="externalLinkDesc" id="externalLinkDescEdit" style="background: var(--base-grey); width: inherit;"></textarea>
+						</div>
+					</div>
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:1918]: indentation

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
- labels all say null now.They should still show the label just with the description  in addition. If there is neither label nor description there should just be now hover.
image
- the public site doesn't show the description.

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
it's not clear what is the label and what the description when it's displayed:
image

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
for external links, the label says "[object Object]"

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditSpaceLinkController.java [hunk 7]
```diff
@@ -33,24 +34,28 @@
...
         }
         byte[] linkImage = null;
         String filename = null;
-        if (file != null) {
+        if (file != null && !file.isEmpty()) {
             linkImage = file.getBytes();
             filename = file.getOriginalFilename();
+        } else if(imageId==null || imageId.equals("")){
+            String[] token = spaceLinkImageIdEdit.split("/");
+            imageId = token[token.length - 1];
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditSpaceLinkController.java:54]: is this part of this ticket?

**Review by jdamerow — CHANGES_REQUESTED** [review]
there seem to be changes not related to this story in here, aren't there?

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditSpaceLinkController.java (hunk 7 above)
Comment (jdamerow) [reply vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditSpaceLinkController.java:54]: ?

**Comment by pradeep-017** [comment]
Hi Julia,
I merged VSPC-172 and VSPC-196 in this ticket and raised a PR. Many UI issues were common and dealing separately would take more time so I merged them together.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditSpaceLinkController.java [hunk 8]
```diff
@@ -33,24 +34,28 @@
...
             return validation;
         }
         byte[] linkImage = null;
         String filename = null;
-        if (file != null) {
+        if (file != null && !file.isEmpty()) {
             linkImage = file.getBytes();
             filename = file.getOriginalFilename();
+        } else if(imageId==null || imageId.equals("")){
+            String[] token = spaceLinkImageIdEdit.split("/");
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditSpaceLinkController.java:53]: Why is this necessary?

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Resolve conflicts please

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java [hunk 9]
```diff
@@ -123,8 +144,20 @@ protected void setDisplayProperties(ILinkDisplay linkDisplay, String id, float p
...
             linkDisplay.setImage(image);
+        } 
+        
+    }
+    
+    protected void setDisplayProperties(ILinkDisplay linkDisplay, float positionX,float positionY,int rotation, DisplayType displayType, String existingImageId) throws ImageCouldNotBeStoredException, ImageDoesNotExistException {
+        linkDisplay.setPositionX(positionX);
+        linkDisplay.setPositionY(positionY);
+        linkDisplay.setRotation(rotation);
+        linkDisplay.setType(displayType != null ? displayType : DisplayType.ARROW);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java:155]: duplicate code; there should be an extra method that's called from both `setDisplayProperties` methods.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddExternalLinkController.java [hunk 10]
```diff
@@ -53,10 +57,9 @@
             filename = file.getOriginalFilename();
         }
         DisplayType type = displayType.isEmpty() ? null : DisplayType.valueOf(displayType);
-        ExternalLinkDisplayMode externalLinkOpenMode = howToOpen.isEmpty() ? null
-                : ExternalLinkDisplayMode.valueOf(howToOpen);
-        IExternalLinkDisplay display = externalLinkManager.createLink(title, id, new Float(x), new Float(y), 0,
-                externalLink, title, type, linkImage, filename, externalLinkOpenMode);
+
+        IExternalLinkDisplay display = externalLinkManager.createLink(title, id, new Float(x), new Float(y), 0, externalLink, title, desc, type, linkImage, filename, imageId);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddExternalLinkController.java:61]: It looks like `externalLinkOpenMode` was dropped here.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditExternalLinkController.java [hunk 11]
```diff
@@ -32,13 +33,15 @@
...
-            @RequestParam("externalLinkIdValueEdit") String externalLinkIdValueEdit,
-            @RequestParam("externalLinkDisplayId") String externalLinkDisplayId,
-            @RequestParam("tabOpen") String howToOpen, @RequestParam("type") String displayType,
-            @RequestParam("externalLinkImage") MultipartFile file) throws SpaceDoesNotExistException, IOException,
-            LinkDoesNotExistsException, NumberFormatException, ImageCouldNotBeStoredException {
+            @RequestParam("externalLinkIdValueEdit") String externalLinkIdValueEdit, @RequestParam("externalLinkDisplayId") String externalLinkDisplayId,
+            @RequestParam(value="externalLinkImageIdEdit", required=false) String externalLinkImageIdEdit,
+            @RequestParam("type") String displayType, @RequestParam(value="externalLinkImage", required=false) MultipartFile file,
+            @RequestParam(value="imageId", required=false) String imageId)
+                    throws SpaceDoesNotExistException, IOException, LinkDoesNotExistsException, NumberFormatException, ImageCouldNotBeStoredException, ImageDoesNotExistException {
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditExternalLinkController.java:43]: `howToOpen` is missing I believe

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditExternalLinkController.java [hunk 12]
```diff
@@ -51,14 +54,10 @@
...
-                : ExternalLinkDisplayMode.valueOf(howToOpen);
-        IExternalLinkDisplay display = (IExternalLinkDisplay) externalLinkManager.updateLink(title, id, new Float(x),
-                new Float(y), 0, externalLink, title, externalLinkIdValueEdit, externalLinkDisplayId, type, linkImage,
-                filename, externalLinkOpenMode);
-        return success(display.getExternalLink().getId(), display.getId(), display.getPositionX(),
-                display.getPositionY(), display.getRotation(), display.getExternalLink().getExternalLink(), title,
-                displayType, null, null);
+
+        IExternalLinkDisplay display = (IExternalLinkDisplay) externalLinkManager.updateLink(title, id, new Float(x), new Float(y), 0, externalLink, title, desc, externalLinkIdValueEdit, externalLinkDisplayId, type, linkImage, filename, imageId);        
+        return success(display.getExternalLink().getId(), display.getId(), display.getPositionX(), display.getPositionY(), display.getRotation(), display.getExternalLink().getExternalLink(), title,displayType,null,null);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditExternalLinkController.java:59]: same here, `howToOpen` is missing

**Review by jdamerow — CHANGES_REQUESTED** [review]
I'll stop my review here. It looks like you just accepted your changes over the incoming ones which means you're reverting the code to a previous version (e.g. the changes to open an external link in a new window). Please go through the code carefully and make sure you keep all new changes coming in from develop.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddExternalLinkController.java (hunk 10 above)
Comment (bnayak4) [reply vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddExternalLinkController.java:61]: @jdamerow for the new changes externalLinkOpenMode is not being used only
